chore: add alert for mac podman socket failures

cdrage opened the pull request:
chore: add alert for mac podman socket failures

### What does this PR do?

* Adds an error message when we've detected a non-disguised socket when
  the "Enable" button has already been pressed
* This is only checked on initial startup and does not perform any
  checks when restarting the machine, etc. This is intentional (for
  now).
* Only on Mac (since we use docker compatibility exclusively there).

### Screenshot / video of UI

<!-- If this PR is changing UI, please include
screenshots or screencasts showing the difference -->

### What issues does this PR fix or reference?

<!-- Include any related issues from Podman Desktop
repository (or from another issue tracker). -->

Closes https://github.com/podman-desktop/podman-desktop/issues/11429

### How to test this PR?

<!-- Please explain steps to verify the functionality,
do not forget to provide unit/component tests -->

- [X] Tests are covering the bug fix or the new feature

1. Press "Enable docker compatibility" in PD.
2. Close PD
3. Open Docker Desktop + let it take over the socket
4. Open PD, should get the error notification on startup due to the
   socket "not being disguised".

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

benoitf commented:
hello, I'm not sure it should be a popup, I expect to be a notification in the dashboard like "Podman needs to be setup"

so you see the message when you're in the dashboard but it's far less intrusive

cdrage commented:
> hello, I'm not sure it should be a popup, I expect to be a notification in the dashboard like "Podman needs to be setup"
> 
> so you see the message when you're in the dashboard but it's far less intrusive

My mistake! I misread the original issue. I'll change it to task manager error / rather than showError.

cdrage commented:
So need your opinion @Firewall 

In the issue it says to use "task manager notification".

I encountered two problems where we're unable to do a proper error reporting, see:

https://github.com/podman-desktop/podman-desktop/issues/11900
https://github.com/podman-desktop/podman-desktop/issues/11899

So the result when doing:

    extensionApi.window.showNotification({
      title: `Compatibility mode error`,
      body: `The Docker socket is not disguised as a Podman socket. Please check any tools that may be overriding /var/run/docker.sock and re-enable compatibility mode.`, 
      type: 'error',
      highlight: true,
    })
    
Results in:
![Screenshot 2025-03-28 at 8 54 47 AM](https://github.com/user-attachments/assets/b623c9dd-c9f3-490d-a452-eed9fa1d3d50)

Which does not give any info / not even appearing as an error.

So unfortunatley we have to choose between:


1. Normal error notification:

![image](https://github.com/user-attachments/assets/29fe3c92-47d1-4319-ac0d-d417666eb60a)


2. Similar notification to podman-mac-helper:

![Screenshot 2025-03-28 at 8 57 48 AM](https://github.com/user-attachments/assets/a92b2457-fa18-42a5-b346-2d27f9af5f3e)

3. Or we can work on issues:

https://github.com/podman-desktop/podman-desktop/issues/11900
https://github.com/podman-desktop/podman-desktop/issues/11899

And have it in task manager like you asked. But we'll have to fix the two above issues first.

Let me know what we should do.

Firewall commented:
I would go with `2. Similar notification to podman-mac-helper:` then. The normal notification error gives it a little bit too much priority because it totally interupts everything, I don't think this notification is that important.

Firewall commented:
But! the message from the popup/dialog was much better. Can you update it to that one (`Podman isn't running in Docker compatibility...`)

cdrage commented:
@benoitf @Firewall 

I've updated it so the notification is now on the dashboard (rather than task manager as discussed above) and updated the PR!

Ready for review 🚀 

![Screenshot 2025-03-31 at 9 55 02 AM](https://github.com/user-attachments/assets/9d4ee4ef-9501-420c-92d1-8eeaabb21bc7)

Firewall commented:
Hey @cdrage, I checkout out the branch and the error message shows up correctly. Awesome! I think there is some padding missing around the text on the button. You can see it on your screenshot too. 

Other than that, I really like the use of the word "disguised" as it explains well what we are doing. However on the page we take the users for the solution we don't really use that word. I have a feeling users are going to click on the button and then be lost which setting to change on that page. WDYT about changing the text on the Docker compatibility for the `Third-Party Docker Tool Compatibility` to 

```
On startup try to disguise the Docker socket (/var/run/docker.sock) with Podman. This will allow tools that rely on this socket to work with Podman. 
```
and also keeping the conditional text to enable/disable like there is now.
```
Disabling this option will prevent third-party Docker-compatible tools to use Podman. [Disable]
```

cdrage commented:
> Hey @cdrage, I checkout out the branch and the error message shows up correctly. Awesome! I think there is some padding missing left&right around the text on the button. You can see it on your screenshot too. ![image](https://private-user-images.githubusercontent.com/505000/429468658-55d4cc3e-5ab0-42c8-b7bf-8cd89b4b1d16.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM2MjIyOTQsIm5iZiI6MTc0MzYyMTk5NCwicGF0aCI6Ii81MDUwMDAvNDI5NDY4NjU4LTU1ZDRjYzNlLTVhYjAtNDJjOC1iN2JmLThjZDg5YjRiMWQxNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQwMlQxOTI2MzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MmU5ZmMwZWI3MmJmYWUxYTgwNjc4ODc2YTc4NjU1ODQ2NjRkYTU3ZjMzYWJiMGNjNTRmMTQ3NDllOTQwN2UzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ZC0e5NQtuacIt7oSpa43mQojZ0nB5DE8ArNsJEN5tns)
> 
> Other than that, I really like the use of the word "disguised" as it explains well what we are doing. However on the page we take the users for the solution we don't really use that word. I have a feeling users are going to click on the button and then be lost which setting to change on that page. WDYT about changing the text on the Docker compatibility for the `Third-Party Docker Tool Compatibility` to
> 
> ```
> On startup try to disguise the Docker socket (/var/run/docker.sock) with Podman. This will allow tools that rely on this socket to work with Podman. 
> ```
> 
> and also keeping the conditional text to enable/disable like there is now.
> 
> ```
> Disabling this option will prevent third-party Docker-compatible tools to use Podman. [Disable]
> ```

I agree on both!

I've opened up issues for both. The first is a minor UI/UX bug: https://github.com/podman-desktop/podman-desktop/issues/11994

With regards to updating the other section, that'd be outside the scope of this PR, but I've opened up https://github.com/podman-desktop/podman-desktop/issues/11995 to get other people's opinions on it too.

feloy reviewed (approved):
LGTM codewise

Firewall commented:
> I agree on both!
> 
> I've opened up issues for both. The first is a minor UI/UX bug: #11994
> 
> With regards to updating the other section, that'd be outside the scope of this PR, but I've opened up #11995 to get other people's opinions on it too.

Ok 👍

cdrage commented:
Rebased and ready for re-review!

ping @axel7083

benoitf reviewed (changes requested):
I started a podman machine while having something using the socket 

And I get for that a warning in the CLI:

```
Starting machine "another"

This machine is currently configured in rootless mode. If your containers
require root permissions (e.g. ports < 1024), or if you run into compatibility
issues with non-podman clients, you can switch using the following command:

	podman machine set --rootful another

API forwarding listening on: /var/folders/_s/rlbw3qg15vx0j6vr4l7kd1hm0000gn/T/podman/another-api.sock

Another process was listening on the default Docker API socket address.
You can still connect Docker API clients by setting DOCKER_HOST using the
following command in your terminal session:

        export DOCKER_HOST='unix:///var/folders/_s/rlbw3qg15vx0j6vr4l7kd1hm0000gn/T/podman/another-api.sock'

Machine "another" started successfully
```

but no warning in Podman Desktop

benoitf commented:
note: I had to disable and re-enable the podman extension to be notified

benoitf commented:
ok but then at extension startup it's also reporting false messages

I fixed the issue and I still have the message so as a user I would think it's not working

benoitf commented:
basically two worklows will fail and this is what people might do:

1. start first podman desktop and then docker desktop: in this case: no notification while you expect to see one.
2. stop docker desktop and restart your podman machine: notification can still be there while you fixed the issue

instead of doing it every 5s, you can run it after a container engine provider has been added or removed

cdrage commented:
> basically two worklows will fail and this is what people might do:
> 
> 1. start first podman desktop and then docker desktop: in this case: no notification while you expect to see one.
> 2. stop docker desktop and restart your podman machine: notification can still be there while you fixed the issue
> 
> instead of doing it every 5s, you can run it after a container engine provider has been added or removed

I've updated the code to check and notify when doing `doUpdateMachines` https://github.com/cdrage/podman-desktop/blob/56d9714acb27156569731c7c429d7f4dfdfb548b/extensions/podman/packages/extension/src/extension.ts#L261-L262

Similar to the podman mac setup mac helper alert).

Can you please try again and see if it gets dismissed on your podman machine restart / no issues?

The only problem is that SOMETIMES our check fails (https://github.com/cdrage/podman-desktop/blob/56d9714acb27156569731c7c429d7f4dfdfb548b/extensions/podman/packages/extension/src/utils/warnings.ts#L29-L30) but that's outside the scope of this PR. 

We need some better compatibility checks in the future.

cdrage commented:
> basically two worklows will fail and this is what people might do:
> 
> 1. start first podman desktop and then docker desktop: in this case: no notification while you expect to see one.
> 2. stop docker desktop and restart your podman machine: notification can still be there while you fixed the issue
> 
> instead of doing it every 5s, you can run it after a container engine provider has been added or removed

I've updated the code to check and notify when doing `doUpdateMachines` https://github.com/cdrage/podman-desktop/blob/56d9714acb27156569731c7c429d7f4dfdfb548b/extensions/podman/packages/extension/src/extension.ts#L261-L262

Similar to the podman mac setup mac helper alert).

Can you please try again and see if it gets dismissed on your podman machine restart / no issues?

The code is based on https://github.com/cdrage/podman-desktop/blob/56d9714acb27156569731c7c429d7f4dfdfb548b/extensions/podman/packages/extension/src/utils/warnings.ts#L29-L30 which checks if it's "disguised" or not.

Whether or not this needs improvement in the future I'm unsure of. I was able to have the notification dismissed on restart on my machine, but I did get some false positives in the past I'm unable to replicate again.

benoitf reviewed (approved):
now it has detected the case

for one, it seems it took a very long time to detect but at least it was ok

cdrage commented:
Enabling auto-merge as two LGTM's and confirmed passing tests.
